x/relationships/keeper/msg_server.go

```diff
@@ -0,0 +1,165 @@
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "relationship from %s to %s already exists inside subspace %d",
+			msg.Signer, msg.Counterparty, msg.SubspaceID)
+	}
+
+	// Save the relationship
+	k.SaveRelationship(ctx, types.NewRelationship(msg.Signer, msg.Counterparty, msg.SubspaceID))
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.Signer),
+		),
+		sdk.NewEvent(
+			types.EventTypeRelationshipCreated,
+			sdk.NewAttribute(types.AttributeRelationshipCreator, msg.Signer),
+			sdk.NewAttribute(types.AttributeRelationshipCounterparty, msg.Counterparty),
+			sdk.NewAttribute(types.AttributeKeySubspace, fmt.Sprintf("%d", msg.SubspaceID)),
+		),
+	})
+
+	return &types.MsgCreateRelationshipResponse{}, nil
+}
+
+// DeleteRelationship defines a rpc method for MsgDeleteRelationship
+func (k msgServer) DeleteRelationship(goCtx context.Context, msg *types.MsgDeleteRelationship) (*types.MsgDeleteRelationshipResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Check if the subspace exists
+	if !k.DoesSubspaceExist(ctx, msg.SubspaceID) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "subspace with id %d not found", msg.SubspaceID)
+	}
+
+	// Check if the relationship exists
+	if !k.HasRelationship(ctx, msg.Signer, msg.Counterparty, msg.SubspaceID) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "relationship from %s to %s does not exist inside subspace %d",
+			msg.Signer, msg.Counterparty, msg.SubspaceID)
+	}
+
+	k.Keeper.DeleteRelationship(ctx, msg.Signer, msg.Counterparty, msg.SubspaceID)
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.Signer),
+		),
+		sdk.NewEvent(
+			types.EventTypeRelationshipsDeleted,
+			sdk.NewAttribute(types.AttributeRelationshipCreator, msg.Signer),
+			sdk.NewAttribute(types.AttributeRelationshipCounterparty, msg.Counterparty),
+			sdk.NewAttribute(types.AttributeKeySubspace, fmt.Sprintf("%d", msg.SubspaceID)),
+		),
+	})
+
+	return &types.MsgDeleteRelationshipResponse{}, nil
+}
+
+// BlockUser defines a rpc method for MsgBlockUser
+func (k msgServer) BlockUser(goCtx context.Context, msg *types.MsgBlockUser) (*types.MsgBlockUserResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Check if the subspace exists
+	if !k.DoesSubspaceExist(ctx, msg.SubspaceID) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "subspace with id %d not found", msg.SubspaceID)
+	}
+
+	// Check if the receiver has blocked the sender before
+	if k.HasUserBlocked(ctx, msg.Blocker, msg.Blocked, msg.SubspaceID) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "%s has already blocked %s", msg.Blocker, msg.Blocked)
+	}
+
```

Comment:
Shouldn't we also check if the two users has a relationship and, in that case, delete that as well?

Reply:
I think it's best to leave DApp handle this case where a relationship and a block exist between two users. One DApp might want to preserve the relationships in the case the block gets removed after. Once other might want to deleted it entirely. I think it's best to leave them there